### PR TITLE
Add specific clear() function for `CompImageHeader` class

### LINKED ISSUE #3526: io.fits: strange behaviour of CompImageHDU.header.clear()
When executing this snippet:

``` Python
In [1]: from astropy.io import fits
In [2]: hdulist = fits.HDUList([fits.PrimaryHDU(), fits.CompImageHDU()])
In [3]: hdulist[1].header['FOO'] = 'BAR'
In [4]: hdulist[1].header.clear()
In [5]: hdulist.writeto('/tmp/tmp.fits', clobber=True)
```

... the header of `hdulist[1]` appears to be empty (as expected):

``` Python
In [6]: len(hdulist[1].header)
Out[6]: 0
```

... but the file that was written to disk pretends that the call to `hdulist[1].header.clear()` never happened:

``` Python
$ fitsheader -e1 /tmp/tmp.fits 
# HDU 1 in /tmp/tmp.fits:
XTENSION= 'IMAGE   '           / Image extension                                
BITPIX  =                    8 / data type of original image                    
NAXIS   =                    0 / dimension of original image                    
PCOUNT  =                    0 / number of parameters                           
GCOUNT  =                    1 / number of groups                               
FOO     = 'BAR     '
```

In contrast, when running this example on a non-compressed ImageHDU, an exception is raised as expected:

``` Python
In [7]: from astropy.io import fits
In [8]: hdulist = fits.HDUList([fits.PrimaryHDU(), fits.ImageHDU()])
In [9]: hdulist[1].header['FOO'] = 'BAR'
In [10]: hdulist[1].header.clear()
In [11]: hdulist.writeto('/tmp/tmp.fits', clobber=True)
---------------------------------------------------------------------------
VerifyError                               Traceback (most recent call last)
<ipython-input-11-755c6620d1e4> in <module>()
----> 1 hdulist.writeto('/tmp/tmp.fits', clobber=True)

/local/home/gb/dev/astropy/astropy/io/fits/hdu/hdulist.py in writeto(self, fileobj, output_verify, clobber, checksum)                                                                         
    660             return                                                                                                                                                                    
    661                                                                                                                                                                                       
--> 662         self.verify(option=output_verify)                                                                                                                                             
    663                                                                                                                                                                                       
    664         # make sure the EXTEND keyword is there if there is extension                                                                                                                 

/local/home/gb/dev/astropy/astropy/io/fits/verify.py in verify(self, option)                                                                                                                  
    118                     warnings.warn(line, VerifyWarning)                                                                                                                                
    119             else:                                                                                                                                                                     
--> 120                 raise VerifyError('\n' + '\n'.join(messages))                                                                                                                         
    121                                                                                                                                                                                       
    122                                                                                                                                                                                       

VerifyError: 
Verification reported errors:
HDU 1:
    'XTENSION' card does not exist.
    'BITPIX' card at the wrong place (card 0).
    'NAXIS' card at the wrong place (card 1).
    'PCOUNT' card does not exist.
    'GCOUNT' card does not exist.
    'PCOUNT' card does not exist.
Note: PyFITS uses zero-based indexing.
```

Could this be a residual bug related to #2172?

### PR DESCRIPTION
**Pull request by cmarmo**
### Description

This pull request adds the `clear()` function for `CompImageHDU` class.

It fixes #3526.

Tests are added using the code snippet in the issue.

### Checklist for package maintainer(s)

This checklist is meant to remind the package maintainer(s) who will review this pull request of some common things to look for. This list is not exhaustive.

- [ ] Do the proposed changes actually accomplish desired goals?
- [ ] Do the proposed changes follow the [Astropy coding guidelines](https://docs.astropy.org/en/latest/development/codeguide.html)?
- [ ] Are tests added/updated as required? If so, do they follow the [Astropy testing guidelines](https://docs.astropy.org/en/latest/development/testguide.html)?
- [ ] Are docs added/updated as required? If so, do they follow the [Astropy documentation guidelines](https://docs.astropy.org/en/latest/development/docguide.html#astropy-documentation-rules-and-guidelines)?
- [ ] Is rebase and/or squash necessary? If so, please provide the author with appropriate instructions. Also see ["When to rebase and squash commits"](https://docs.astropy.org/en/latest/development/when_to_rebase.html).
- [ ] Did the CI pass? If no, are the failures related? If you need to run daily and weekly cron jobs as part of the PR, please apply the `Extra CI` label.
- [ ] Is a change log needed? If yes, did the change log check pass? If no, add the `no-changelog-entry-needed` label. If this is a manual backport, use the `skip-changelog-checks` label unless special changelog handling is necessary.
- [ ] Is this a big PR that makes a "What's new?" entry worthwhile and if so, is (1) a "what's new" entry included in this PR and (2) the "whatsnew-needed" label applied?
- [ ] Is a milestone set? Milestone must be set but `astropy-bot` check might be missing; do not let the green checkmark fool you.
- [ ] At the time of adding the milestone, if the milestone set requires a backport to release branch(es), apply the appropriate `backport-X.Y.x` label(s) *before* merge.

### CODE REVIEW

**Comment by cmarmo** [comment]
I force-pushed to check the documentation failure. Everything is fine now. Thanks for your understanding.

File: astropy/io/fits/hdu/compressed.py [hunk 1]
```diff
@@ -374,6 +375,16 @@ def _remap_index(self, idx):
...
         return idx
 
+    def clear(self):
+        """
+        Remove all cards from the header.
+        """
+
+        self._table_header._cards = []
+        self._table_header._keyword_indices = collections.defaultdict(list)
+        self._table_header._rvkc_indices = collections.defaultdict(list)
```
Comment (saimn) [inline astropy/io/fits/hdu/compressed.py:385]: This should probably work ?
```suggestion
        self._table_header.clear()
```

File: astropy/io/fits/hdu/table.py [hunk 2]
```diff
@@ -537,27 +537,28 @@ def _verify(self, option='warn'):
...
+
+            self.req_cards('NAXIS', None, lambda v: (v == 2), 2, option, errs)
+            self.req_cards('BITPIX', None, lambda v: (v == 8), 8, option, errs)
+            self.req_cards('TFIELDS', 7,
+                           lambda v: (_is_int(v) and v >= 0 and v <= 999), 0,
+                           option, errs)
+            tfields = self._header['TFIELDS']
+            for idx in range(tfields):
+                self.req_cards('TFORM' + str(idx + 1), None, None, None, option,
+                               errs)
```
Comment (saimn) [inline astropy/io/fits/hdu/table.py:561]: I don't understand why this change ? (It seems to give the same error in the end).

File: docs/changes/io.fits/13102.bugfix.rst [hunk 3]
```diff
@@ -0,0 +1 @@
+Add specific ``clear()`` function for ``CompImageHeader`` class.
```
Comment (saimn) [inline docs/changes/io.fits/13102.bugfix.rst:1]:
```suggestion
Fix ``CompImageHeader.clear()``.
```

**Review by saimn — COMMENTED** [review]
Thanks for addressing those old issues @cmarmo :). A few comments below.

File: astropy/io/fits/hdu/compressed.py (hunk 1 above)
Comment (cmarmo) [reply astropy/io/fits/hdu/compressed.py:385]: Yes, indeed, sorry for that... I guess I messed up in my debugging back and forth.

File: astropy/io/fits/hdu/table.py (hunk 2 above)
Comment (cmarmo) [reply astropy/io/fits/hdu/table.py:561]: I have open two fake PRs in my repo to explain this.
The [first one](https://github.com/cmarmo/astropy/pull/2) only contains the new ``clear()`` function.
The specific test is [failing](https://github.com/cmarmo/astropy/runs/6085659937?check_suite_focus=true#step:6:1455).
The [second PR](https://github.com/cmarmo/astropy/pull/3) adds the control on the header length and [passes](https://github.com/cmarmo/astropy/runs/6085908385?check_suite_focus=true#step:6:1911).
Also the checks on the table specific keywords should be in the `if` compound otherwise the test fails with a different error.
My understanding is that ``CompImageHDU`` inherits from ``BinTableHDU`` and this forces the header to be verified in two ``verify()`` functions while ImageHDU headers are verified only once.... but I know I have a partial understanding of the whole picture....

**Review by saimn — APPROVED** [review]
LGTM, many thanks @cmarmo for thanks for tackling those old issues!

**Comment by pllim** [comment]
@meeseeksdev backport to v5.0.x
